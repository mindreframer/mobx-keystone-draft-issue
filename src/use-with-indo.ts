import { observable } from "mobx";
import { Draft, draft, model, Model, modelAction, prop } from "mobx-keystone";

@model("myApp/Tag")
export class Tag extends Model({ label: prop<string>("").withSetter() }) {}
@model("myApp/Form1")
export class Form1 extends Model({
  tags: prop<Tag[]>(() => []).withSetter(),
  name: prop<string>("").withSetter(),
}) {}

@model("myApp/Form2")
export class Form2 extends Model({
  tags: prop<Tag[]>(() => []).withSetter(),
  name: prop<string>("").withSetter(),
}) {}

@model("myApp/Settings")
export class Settings extends Model({
  name: prop<string>("").withSetter(),
  form1List: prop<Form1[]>(() => []).withSetter(),
  form2List: prop<Form2[]>(() => []).withSetter(),
  tags: prop<Tag[]>(() => []),
}) {
  @modelAction arrayRemoveAt(array: any[], index: number) {
    array.splice(index, 1);
  }

  @modelAction arrayPush(array: any[], item: any) {
    array.push(item);
  }

  @modelAction addForm1() {
    this.form1List.push(new Form1({}));
  }

  @modelAction addForm2() {
    this.form2List.push(new Form2({}));
  }
}

@model("myApp/Root")
export class Root extends Model({
  settings: prop<Settings | undefined>(),
}) {
  @observable.ref settingsDraft: Draft<Settings> | undefined;
  onInit() {
    if (this.settings) {
      this.setSettings(this.settings);
    }
  }
  @modelAction setSettings(val: Settings) {
    this.settings = val;
    this.settingsDraft = draft(this.settings);
  }
  @modelAction commit() {
    this.settingsDraft?.commit();
  }

  @modelAction deleteTag(index: number) {
    this.settingsDraft?.data.tags.splice(index, 1);
  }
}
