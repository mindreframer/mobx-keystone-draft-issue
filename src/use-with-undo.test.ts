import { expect } from "@jest/globals";
import { toJS } from "mobx";
import { undoMiddleware } from "mobx-keystone";
import { Root, Settings, Tag } from "./use-with-indo";
import "./commonSetup";

const logAll = (v: any, label = "") => {
  console.log(label, JSON.stringify(toJS(v), null, 2));
};

describe("Root", () => {
  it("should support drafts wrapped in undo/redo middleware ", () => {
    const root = new Root({});
    root.setSettings(new Settings({}));
    const settings = new Settings({
      tags: [new Tag({ label: "tag1" }), new Tag({ label: "tag2" })],
    });
    root.setSettings(settings);
    const undoManagerSettingsDraft = undoMiddleware(root.settingsDraft!.data);
    const draftModel = root.settingsDraft!;
    draftModel.data.arrayRemoveAt(draftModel.data.tags, 0); // this is recorded
    expect(draftModel).toMatchSnapshot("DRAFT after direct modification");
    undoManagerSettingsDraft.undo();
    expect(draftModel.data.tags.length).toBe(2);
    undoManagerSettingsDraft.redo();
    expect(draftModel.data.tags.length).toBe(1);
    draftModel.data.setName("HELLO");

    expect(undoManagerSettingsDraft.store.undoEvents).toMatchSnapshot(
      "Undo Events"
    );
    root.commit();
    expect(draftModel).toMatchSnapshot("DRAFT AFTER COMMIT");

    draftModel.data.addForm1();
    draftModel.data.addForm1();
    draftModel.data.addForm2();
    draftModel.data.addForm2();

    expect(undoManagerSettingsDraft.store.undoEvents).toMatchSnapshot(
      "Undo Events with forms"
    );

    draftModel.data.arrayPush(
      draftModel.data.form1List[0].tags,
      new Tag({ label: "tag1" })
    );
    draftModel.data.arrayPush(
      draftModel.data.form1List[0].tags,
      new Tag({ label: "tag2" })
    );

    expect(draftModel.data.form1List[0].tags).toMatchSnapshot(
      "tags form1 - 0 - 2 tags added"
    );
    undoManagerSettingsDraft.undo();
    expect(draftModel.data.form1List[0].tags).toMatchSnapshot(
      "tags form1 - 0 - undo -> 1 tag left"
    );

    undoManagerSettingsDraft.redo();

    expect(draftModel.data.form1List[0].tags).toMatchSnapshot(
      "tags form1 - 0 - redo -> 2 tags again"
    );
  });
});
