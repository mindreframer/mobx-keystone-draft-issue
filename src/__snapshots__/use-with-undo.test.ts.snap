// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root should support drafts wrapped in undo/redo middleware : DRAFT AFTER COMMIT 1`] = `
Draft {
  "data": Settings {
    "$": Object {
      "$modelId": "id-5",
      "form1List": Array [],
      "form2List": Array [],
      "name": "HELLO",
      "tags": Array [
        Tag {
          "$": Object {
            "$modelId": "id-4",
            "label": "tag2",
          },
          "$modelType": "myApp/Tag",
        },
      ],
    },
    "$modelType": "myApp/Settings",
  },
  "originalData": Settings {
    "$": Object {
      "$modelId": "id-5",
      "form1List": Array [],
      "form2List": Array [],
      "name": "HELLO",
      "tags": Array [
        Tag {
          "$": Object {
            "$modelId": "id-4",
            "label": "tag2",
          },
          "$modelType": "myApp/Tag",
        },
      ],
    },
    "$modelType": "myApp/Settings",
  },
}
`;

exports[`Root should support drafts wrapped in undo/redo middleware : DRAFT after direct modification 1`] = `
Draft {
  "data": Settings {
    "$": Object {
      "$modelId": "id-5",
      "form1List": Array [],
      "form2List": Array [],
      "name": "",
      "tags": Array [
        Tag {
          "$": Object {
            "$modelId": "id-4",
            "label": "tag2",
          },
          "$modelType": "myApp/Tag",
        },
      ],
    },
    "$modelType": "myApp/Settings",
  },
  "originalData": Settings {
    "$": Object {
      "$modelId": "id-5",
      "form1List": Array [],
      "form2List": Array [],
      "name": "",
      "tags": Array [
        Tag {
          "$": Object {
            "$modelId": "id-3",
            "label": "tag1",
          },
          "$modelType": "myApp/Tag",
        },
        Tag {
          "$": Object {
            "$modelId": "id-4",
            "label": "tag2",
          },
          "$modelType": "myApp/Tag",
        },
      ],
    },
    "$modelType": "myApp/Settings",
  },
}
`;

exports[`Root should support drafts wrapped in undo/redo middleware : Undo Events 1`] = `
Array [
  Object {
    "actionName": "arrayRemoveAt",
    "inversePatches": Array [
      Object {
        "op": "add",
        "path": Array [
          "tags",
          0,
        ],
        "value": Object {
          "$modelId": "id-3",
          "$modelType": "myApp/Tag",
          "label": "tag1",
        },
      },
    ],
    "patches": Array [
      Object {
        "op": "remove",
        "path": Array [
          "tags",
          0,
        ],
      },
    ],
    "targetPath": Array [],
    "type": "single",
  },
  Object {
    "actionName": "setName",
    "inversePatches": Array [
      Object {
        "op": "replace",
        "path": Array [
          "name",
        ],
        "value": "",
      },
    ],
    "patches": Array [
      Object {
        "op": "replace",
        "path": Array [
          "name",
        ],
        "value": "HELLO",
      },
    ],
    "targetPath": Array [],
    "type": "single",
  },
]
`;

exports[`Root should support drafts wrapped in undo/redo middleware : Undo Events with forms 1`] = `
Array [
  Object {
    "actionName": "arrayRemoveAt",
    "inversePatches": Array [
      Object {
        "op": "add",
        "path": Array [
          "tags",
          0,
        ],
        "value": Object {
          "$modelId": "id-3",
          "$modelType": "myApp/Tag",
          "label": "tag1",
        },
      },
    ],
    "patches": Array [
      Object {
        "op": "remove",
        "path": Array [
          "tags",
          0,
        ],
      },
    ],
    "targetPath": Array [],
    "type": "single",
  },
  Object {
    "actionName": "setName",
    "inversePatches": Array [
      Object {
        "op": "replace",
        "path": Array [
          "name",
        ],
        "value": "",
      },
    ],
    "patches": Array [
      Object {
        "op": "replace",
        "path": Array [
          "name",
        ],
        "value": "HELLO",
      },
    ],
    "targetPath": Array [],
    "type": "single",
  },
  Object {
    "actionName": "addForm1",
    "inversePatches": Array [
      Object {
        "op": "replace",
        "path": Array [
          "form1List",
          "length",
        ],
        "value": 0,
      },
    ],
    "patches": Array [
      Object {
        "op": "add",
        "path": Array [
          "form1List",
          0,
        ],
        "value": Object {
          "$modelId": "id-7",
          "$modelType": "myApp/Form1",
          "name": "",
          "tags": Array [],
        },
      },
    ],
    "targetPath": Array [],
    "type": "single",
  },
  Object {
    "actionName": "addForm1",
    "inversePatches": Array [
      Object {
        "op": "replace",
        "path": Array [
          "form1List",
          "length",
        ],
        "value": 1,
      },
    ],
    "patches": Array [
      Object {
        "op": "add",
        "path": Array [
          "form1List",
          1,
        ],
        "value": Object {
          "$modelId": "id-8",
          "$modelType": "myApp/Form1",
          "name": "",
          "tags": Array [],
        },
      },
    ],
    "targetPath": Array [],
    "type": "single",
  },
  Object {
    "actionName": "addForm2",
    "inversePatches": Array [
      Object {
        "op": "replace",
        "path": Array [
          "form2List",
          "length",
        ],
        "value": 0,
      },
    ],
    "patches": Array [
      Object {
        "op": "add",
        "path": Array [
          "form2List",
          0,
        ],
        "value": Object {
          "$modelId": "id-9",
          "$modelType": "myApp/Form2",
          "name": "",
          "tags": Array [],
        },
      },
    ],
    "targetPath": Array [],
    "type": "single",
  },
  Object {
    "actionName": "addForm2",
    "inversePatches": Array [
      Object {
        "op": "replace",
        "path": Array [
          "form2List",
          "length",
        ],
        "value": 1,
      },
    ],
    "patches": Array [
      Object {
        "op": "add",
        "path": Array [
          "form2List",
          1,
        ],
        "value": Object {
          "$modelId": "id-10",
          "$modelType": "myApp/Form2",
          "name": "",
          "tags": Array [],
        },
      },
    ],
    "targetPath": Array [],
    "type": "single",
  },
]
`;

exports[`Root should support drafts wrapped in undo/redo middleware : tags form1 - 0 - 2 tags added 1`] = `
Array [
  Tag {
    "$": Object {
      "$modelId": "id-11",
      "label": "tag1",
    },
    "$modelType": "myApp/Tag",
  },
  Tag {
    "$": Object {
      "$modelId": "id-12",
      "label": "tag2",
    },
    "$modelType": "myApp/Tag",
  },
]
`;

exports[`Root should support drafts wrapped in undo/redo middleware : tags form1 - 0 - redo -> 2 tags again 1`] = `
Array [
  Tag {
    "$": Object {
      "$modelId": "id-11",
      "label": "tag1",
    },
    "$modelType": "myApp/Tag",
  },
  Tag {
    "$": Object {
      "$modelId": "id-12",
      "label": "tag2",
    },
    "$modelType": "myApp/Tag",
  },
]
`;

exports[`Root should support drafts wrapped in undo/redo middleware : tags form1 - 0 - undo -> 1 tag left 1`] = `
Array [
  Tag {
    "$": Object {
      "$modelId": "id-11",
      "label": "tag1",
    },
    "$modelType": "myApp/Tag",
  },
]
`;
